No UEFI Shell

miczyg1 opened the pull request:
Necessary fixes for no-Shell builds. Details in commit messages.

miczyg1 committed:
DasharoPayloadPkg/DasharoPayloadPkg.dsc: Guard net defines with NETWORK_ENABLE

When trying to build without EFI Shell, a build issue appears saying that
PcdAllowHttpConnections is not declared in DEC files referenced in INF files
in FDF. It occurred that NetworkPcds.dsc.inc initializes a that PCD, but
modules that use this PCD are only built and included into the image if
NETWORK_ENABLE is TRUE. As a result a PCD was defined, but not a single
module actually existed in the image to use it, i.e. use the NetworkPkg.dec
file. This bug lived silently until EFI Shell was removed, because one of the
EFI Shell module INF files included NetworkPkg.dec.

To solve this, guard the network defines with NETWORK_ENABLE, so that the
PCD is only initialized in case where network is actually enabled.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
DasharoPayloadPkg/PlatformBootManagerLib: Register only existing FV boot options

When updating firmware to a build without UEFI Shell, the settings may be preserved,
including the boot options from previous firmware, which results in the non-existent
UEFI Shell still being presented as an active boot option. To avoid confusion, check
the existence of a file before registering a boot option.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
DasharoPayloadPkg/DasharoPayloadPkg.dsc: Clean up libraries

The UefiDevicePathLib from MdePkg is used everywhere, so no need
to define the same library three times in different places. Removed
redunant entries with UefiDevicePathLib.

Also the FileHandleLib seems to be used by QEMU Q35 target and is not
compiled when Shell is removed. Move it to common section with libraries.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 commented:
Tested on VP3230 without shell, then with shell to see if the Shell option is created back.

Workflows that build all platforms with and without UEFI Shell:

https://github.com/Dasharo/coreboot/actions/runs/13203150358
https://github.com/Dasharo/coreboot/actions/runs/13203158695

SergiiDmytruk commented on DasharoPayloadPkg/Library/PlatformBootManagerLib/PlatformBootManager.c:
```diff
@@ -141,6 +153,8 @@ PlatformRegisterFvBootOption (
   ASSERT_EFI_ERROR (Status);
   FreePool (DevicePath);
 
+  // No need to guard here, because an error would be displayed on the screen
```
`No need to guard here` in front of an `if` is confusing without knowing context of this commit.  Would be clearer to mention file existence explicitly.

miczyg1 replied:
Would `No need to guard against FileExists here` be sufficient? @SergiiDmytruk

SergiiDmytruk replied:
Maybe `Whether boot option's file exists is irrelevant, ...` as `guard against FileExists` sounds a bit weird.

miczyg1 replied:
The whole command has been changed here: https://github.com/Dasharo/edk2/pull/208/commits/d14eb88ce5999a9de706f2dd12bceb87a0c5d119

SergiiDmytruk commented on DasharoPayloadPkg/Library/PlatformBootManagerLib/PlatformBootManager.c:
```diff
@@ -152,9 +166,14 @@ PlatformRegisterFvBootOption (
     ASSERT_EFI_ERROR (Status);
+  } else if (OptionIndex != -1 && !FileExists) {
+    // Option exists but no longer in the image. Remove it from boot option list.
+    Status = EfiBootManagerDeleteLoadOptionVariable (BootOptions[OptionIndex].OptionNumber,
```
This makes function for adding an option to remove an already existing one.

I don't really have a better way of implementing it (need some place which can drop all outdated entries and there might be none), but an update to a comment on `PlatformRegisterFvBootOption()`  to warn about this unexpected behaviour seems warranted.

miczyg1 replied:
I may have made a too big shortcut here.  `PlatformRegisterFvBootOption` and `PlatformUnRegisterFvBootOption` have so much common code... If I would like to be proper about it it would have to look like this:

```
BOOLEAN FvFileExists (EFI_GUID FileGuid)
{
  EFI_DEVICE_PATH_PROTOCOL          *FileDevicePath;
  BOOLEAN                           FileExists;
  EFI_STATUS                        Status;
  
  // Search if given file exists anywhere
  Status = GetFileDevicePathFromAnyFv (
             FileGuid,
             EFI_SECTION_ALL,
             0,
             &FileDevicePath
             );

  FileExists = !EFI_ERROR (Status);

  Return FileExists;
}
```

Then in `PlatformBootManagerAfterConsole`:

```
  if (FvFileExists (PcdGetPtr (PcdShellFile)))  {
    PlatformRegisterFvBootOption (PcdGetPtr (PcdShellFile),
                                L"UEFI Shell",
                                LOAD_OPTION_ACTIVE,
                                FALSE);
  } else {
    PlatformUnRegisterFvBootOption (PcdGetPtr (PcdShellFile),
                                L"UEFI Shell",
                                LOAD_OPTION_ACTIVE);
  }
```

For every FV File boot option we create here. BootManagerMenuApp also has a lot of redundancy and separate functions...

Anyways, right now adding the FV file boot options is broken, because it may add files that do not even exist (as a result FW may contain boot options that no longer should exist, e.g. after update). Let me know WDYT @SergiiDmytruk

SergiiDmytruk replied:
Yeah, I remember wondering about those functions while rebasing. I think we can just merge them into one that does whatever is necessary (they aren't public anyway):
```c
STATIC VOID
SyncFvBootOption (
  BOOLEAN                          Present,
  EFI_GUID                         *FileGuid,
  CHAR16                           *Description,
  BOOLEAN                          BootNow
  )
{
```

(I dropped `Attributes` because it's always `LOAD_OPTION_ACTIVE`.)

Then a single invocation will be enough to align boot options with expectations:
```c
  if (FUMEnabled) {
    NetBootEnabled = TRUE;
  } else {
    VarSize = sizeof (NetBootEnabled);
    Status = gRT->GetVariable (
        DASHARO_VAR_NETWORK_BOOT,
        &gDasharoSystemFeaturesGuid,
        NULL,
        &VarSize,
        &NetBootEnabled
        );
    if (Status == EFI_NOT_FOUND || VarSize != sizeof(NetBootEnabled))
      NetBootEnabled = FixedPcdGetBool(PcdDefaultNetworkBootEnable);
  }

  //
  // Register iPXE
  //
  SyncFvBootOption (NetBootEnabled,
                    PcdGetPtr (PcdiPXEFile),
                    (CHAR16 *) PcdGetPtr(PcdiPXEOptionName),
                    FUMEnabled);
  //
  // Register UEFI Shell
  //
  SyncFvBootOption (TRUE, PcdGetPtr (PcdShellFile), L"UEFI Shell", FALSE);

```

And `DEBUG((DEBUG_INFO, ...)` can be printed by the function to reflect what it was trying to do and what was actually done.

miczyg1 replied:
https://github.com/Dasharo/edk2/pull/208/commits/d14eb88ce5999a9de706f2dd12bceb87a0c5d119

miczyg1 committed:
DasharoPayloadPkg: Reduce FV boot option registering redundancy

The RegisterFvBootOption and UnregisterFvBootOption functions shared
a lot of code. Merge those two functions into a new SyncFvBootOption
which checks for file presence and registers the option accordingly or
removes it if the file does not exist or should be removed explicitly.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

miczyg1 committed:
OvmfPkg/OvmfPkgX64.dsc: Fix build with fast boot

Fast Boot feature added a few PCDs to common Dasharo modules. These
PCDs have to be defined as DynamicDefault in the platform DSC file.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>

SergiiDmytruk commented on DasharoPayloadPkg/Library/PlatformBootManagerLib/PlatformBootManager.c:
```diff
@@ -156,73 +170,21 @@ PlatformRegisterFvBootOption (
                   );
 
-  if (OptionIndex == -1) {
+  if (OptionIndex == -1 && FileExists) {
```
I think this lacks a check for `Remove` to prevent adding an option on removal:
```suggestion
  if (OptionIndex == -1 && FileExists && !Remove) {
```

miczyg1 replied:
Amended the commit: https://github.com/Dasharo/edk2/compare/f01ecabd9dc09421a7bc7309e229fbf5427d5218..a309360e147d1dd932950a06b12a0a594f21ea8b

SergiiDmytruk reviewed (approved):
Looks like DBX file got updated again: https://github.com/microsoft/secureboot_objects/commits/main/PostSignedObjects/DBX/amd64/DBXUpdate.bin, otherwise LGTM.

mkopec commented:
yeah, see https://github.com/Dasharo/edk2/pull/212 , hopefully this will help if DBX gets updated more often

miczyg1 commented:
Merged the DBX update PRs, now all checks should pass.
